src/Features/Core/Portable/Completion/CommitChange.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal sealed class CommitChange
+    {
+        /// <summary>
+        /// The changes to be applied to the document.
+        /// </summary>
+        public ImmutableArray<TextChange> TextChanges { get; }
+
+        /// <summary>
+        /// The new caret position after commit.
+        /// If null then the caret position is not specified.
+        /// </summary>
+        public int? NewPosition { get; }
+
+        /// <summary>
+        /// True if the changes include the commit key.
+        /// </summary>
+        public bool IncludesCommitKey { get; }
+
+        private CommitChange(ImmutableArray<TextChange> textChanges, int? newPosition, bool includesCommit)
```

Comment:
What's the formatting story?  Is the service expected to give you the formatted text changes?  Or oes the controller format things?